std/permutation/poseidon2/gkr.go

```diff
@@ -0,0 +1,388 @@
+	}
+	y := api.Mul(x[0], x[0])
+	y = api.Mul(y, y)
+
+	return y
+}
+
+func (g pow4Gate) Degree() int {
+	return 4
+}
+
+// pow4Gate computes a, b -> a^4 * b
+type pow4TimesGate struct{}
+
+func (g pow4TimesGate) Evaluate(api frontend.API, x ...frontend.Variable) frontend.Variable {
+	if len(x) != 2 {
+		panic("expected 1 input")
+	}
+	y := api.Mul(x[0], x[0])
+	y = api.Mul(y, y)
+
+	return api.Mul(y, x[1])
+}
+
+func (g pow4TimesGate) Degree() int {
+	return 5
+}
+
+type pow2Gate struct{}
+
+func (g pow2Gate) Evaluate(api frontend.API, x ...frontend.Variable) frontend.Variable {
+	if len(x) != 1 {
+		panic("expected 1 input")
+	}
+	return api.Mul(x[0], x[0])
+}
+
+func (g pow2Gate) Degree() int {
+	return 2
+}
+
+type pow2TimesGate struct{}
+
+func (g pow2TimesGate) Evaluate(api frontend.API, x ...frontend.Variable) frontend.Variable {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	return api.Mul(x[0], x[0], x[1])
+}
+
+func (g pow2TimesGate) Degree() int {
+	return 3
+}
+
+// for x1, the partial round gates are identical to full round gates
+// for x2, the partial round gates are just a linear combination
+// TODO @Tabaie eliminate the x2 partial round gates and have the x1 gates depend on i - rf/2 or so previous x1's
+
+// extGate2 applies the external matrix mul, outputting the second element of the result
+type extGate2 struct {
+}
+
+func (g *extGate2) Evaluate(api frontend.API, x ...frontend.Variable) frontend.Variable {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	return api.Add(api.Mul(x[1], 2), x[0])
+}
+
+func (g *extGate2) Degree() int {
+	return 1
+}
+
+// intKeyGate2 applies the internal matrix mul, then adds the round key
+type intKeyGate2 struct {
+	roundKey *big.Int
+}
+
+func (g *intKeyGate2) Evaluate(api frontend.API, x ...frontend.Variable) frontend.Variable {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	return api.Add(api.Mul(x[1], 3), x[0], g.roundKey)
+}
+
+func (g *intKeyGate2) Degree() int {
+	return 1
+}
+
+type extGate struct{}
+
+func (g extGate) Evaluate(api frontend.API, x ...frontend.Variable) frontend.Variable {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	return api.Add(api.Mul(x[0], 2), x[1])
+}
+
+func (g extGate) Degree() int {
+	return 1
+}
+
+type GkrPermutations struct {
+	api  frontend.API
+	ins1 []frontend.Variable
+	ins2 []frontend.Variable
+	outs []frontend.Variable
+}
+
+func NewGkrPermutations(api frontend.API) *GkrPermutations {
+	res := GkrPermutations{
+		api: api,
+	}
+	api.Compiler().Defer(res.finalize)
+	return &res
+}
+
+func (p *GkrPermutations) Permute(a, b frontend.Variable) frontend.Variable {
+	s, err := p.api.Compiler().NewHint(permuteHint, 1, a, b)
+	if err != nil {
+		panic(err)
+	}
+	p.ins1 = append(p.ins1, a)
+	p.ins2 = append(p.ins2, b)
+	p.outs = append(p.outs, s[0])
+	return s[0]
+}
+
+func frToInt(x *frBls12377.Element) *big.Int {
+	var res big.Int
+	x.BigInt(&res)
+	return &res
+}
+
+func (p *GkrPermutations) finalize(api frontend.API) error {
```

Comment:
The method is currently quite difficult to follow - it inlines hash registration, gate registration (and all the helpers to derive names). I would separate the gate registration into a separate function (which can be registered automatically at init). I would define them separately at package-level `init()` function.

Then, it would be nice if we would have gates collected as in methods of the Permutation methods a la `matMulExternal` (similar to `matMulExternalInPlace` but doesn't do in place but rather returns the result) etc.

Reply:
Gate registration is the majority of the work compared to the circuit wiring, so separating them doesn't make much sense to me (in that case we'd have to iterate over the circuit structure once more, resulting in code duplication). I added comments and separated the two of them from the boilerplate GKR stuff. Does it look better now?